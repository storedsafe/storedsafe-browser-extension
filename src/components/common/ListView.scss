@import '../../variables';

.list-view {
  height: 100%;
  overflow: hidden auto;

  &.has-selected {
    overflow: hidden;
  }

  .list-view-item-title {
    padding: $spacing;
    background-color: $color-bg-alt;
    color: $color-fg;
    cursor: pointer;
    user-select: none;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);

    &:hover {
      background-color: $color-bg;
    }
  }

  .list-view-item-content {
    display: none;

    background-color: $color-bg-alt;
    color: $color-fg;
    padding: $spacing;
    overflow: hidden auto;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);
  }

  .list-view-item.selected {
    display: flex;
    flex-direction: column;
    height: 100%;

    .list-view-item-title {
      position: relative;
      padding-left: $spacing * 2;
      border-radius: 0 $border-radius 0 0;
    }

    .list-view-item-title::before {
      position: absolute;
      left: 0;
      top: 50%;
      content: '';
      border: solid $color-primary-0;
      border-width: 2px 0 0 2px;
      display: inline-block;
      width: 6px;
      height: 6px;
      transform: rotate(-45deg);
      margin: -4px 0 0 $spacing - 2.5px;
    }

    .list-view-item-content {
      display: block;
      flex-grow: 1;
      border-radius: 0 0 $border-radius $border-radius;
    }

    .list-view-item-title,
    .list-view-item-content {
      transition: border-radius $transition-duration;
    }
  }

  .list-view-item.hidden {
    .list-view-item-title {
      display: none;
    }
  }
}
