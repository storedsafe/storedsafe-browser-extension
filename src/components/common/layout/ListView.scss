@import '../../../styles/variables.scss';

.list-view {
  height: 100%;
  overflow: hidden auto;

  &.has-selected {
    overflow: hidden;
  }

  .list-view-item-title {
    width: 100%;
    border: 0;
    outline: 0;
    border-bottom: 1px solid $color-bg;
    padding: $spacing;
    text-align: start;
    background-color: $color-bg-alt;
    color: $color-fg;
    cursor: pointer;
    user-select: none;
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.1);

    &:hover {
      background-color: $color-bg;
    }

    &:focus {
      background-color: $color-bg;
      border-bottom: 1px solid $color-accent-0;
    }
  }

  .list-view-item-content {
    display: none;

    background-color: $color-bg-alt;
    color: $color-fg;
    padding: $spacing;
    overflow: hidden auto;
    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);
  }

  .list-view-item.selected {
    display: flex;
    flex-direction: column;
    height: 100%;

    .list-view-item-title {
      position: relative;
      padding-left: $spacing * 2;
    }

    .list-view-item-title::before {
      position: absolute;
      left: 0;
      top: 50%;
      content: '';
      border: solid $color-primary-0;
      border-width: 2px 0 0 2px;
      display: inline-block;
      width: 6px;
      height: 6px;
      transform: rotate(-45deg);
      margin: -4px 0 0 $spacing - 2.5px;
    }

    .list-view-item-content {
      display: block;
      flex-grow: 1;
    }
  }

  .list-view-item.hidden {
    .list-view-item-title {
      display: none;
    }
  }
}
