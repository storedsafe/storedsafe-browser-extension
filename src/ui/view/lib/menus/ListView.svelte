<script lang="ts">
  import type { ListItem } from './ListView'
  import { createEventDispatcher } from "svelte";
  import ListItemBox from "./ListItemBox.svelte";

  const dispatch = createEventDispatcher();

  export let items: ListItem<any>[] = [];
  export let selected: any = null;

  function setSelected(name: any) {
    dispatch("select", selected !== name ? name : null);
  }
</script>

<style>
  section.selected {
    row-gap: 0;
  }
</style>

<section class:selected={!!selected} class="grid">
  {#each items as { component, props, name } (name)}
    {#if !selected || selected === name}
      <ListItemBox on:click={() => setSelected(name)} selected={selected === name}>
        <svelte:component this={component} selected={selected === name} {...props} />
      </ListItemBox>
    {/if}
  {/each}
</section>
