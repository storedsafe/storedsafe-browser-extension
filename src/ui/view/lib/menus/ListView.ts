import type { Component } from "svelte";

export interface ListItem<T extends Record<string, any>> {
  Component: Component<T, any, any>;
  name: any;
  props: T;
}
