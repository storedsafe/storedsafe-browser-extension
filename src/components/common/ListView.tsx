import React, { useState, useEffect } from 'react'
import './ListView.scss'

export interface ListItem {
  key: string
  title: React.ReactNode
  content: React.ReactNode
}

interface ListViewProps {
  items: ListItem[]
  defaultSelected?: string
}

export const ListView: React.FunctionComponent<ListViewProps> = ({
  items,
  defaultSelected
}: ListViewProps) => {
  const [selected, setSelected] = useState<string>(defaultSelected)

  useEffect(() => {
    setSelected(defaultSelected)
  }, [defaultSelected])

  const hasSelected = selected !== undefined

  const list = items.map(item => {
    const isSelected = selected === item.key
    const onSelect = (): void => {
      if (isSelected) {
        setSelected(undefined)
      } else {
        setSelected(item.key)
      }
    }
    return (
      <article
        className={`list-view-item${isSelected ? ' selected' : ''}${
          hasSelected && !isSelected ? ' hidden' : ''
        }`}
        key={item.key}
      >
        <article className='list-view-item-title' onClick={onSelect}>
          {item.title}
        </article>
        <article className='list-view-item-content'>{item.content}</article>
      </article>
    )
  })

  return (
    <section className={`list-view${hasSelected ? ' has-selected' : ''}`}>
      {list}
    </section>
  )
}
