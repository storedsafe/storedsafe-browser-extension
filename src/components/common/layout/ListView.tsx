import React, { useState, useEffect } from 'react'
import './ListView.scss'

export interface ListItem {
  key: string
  title: React.ReactNode
  content: React.ReactNode
}

interface ListViewProps {
  items: ListItem[]
  defaultSelected?: string
}

export const ListView: React.FunctionComponent<ListViewProps> = ({
  items,
  defaultSelected
}: ListViewProps) => {
  const defaultExists = items.findIndex(({ key }) => defaultSelected) !== -1
  const [selected, setSelected] = useState<string>(defaultExists ? defaultSelected : undefined)

  useEffect(() => {
    const newDefaultExists = items.findIndex(({ key }) => defaultSelected) !== -1
    setSelected(newDefaultExists ? defaultSelected : undefined)
  }, [defaultSelected])

  const hasSelected = selected !== undefined

  const list = items.map(item => {
    const isSelected = selected === item.key
    const onSelect = (): void => {
      if (isSelected) {
        setSelected(undefined)
      } else {
        setSelected(item.key)
      }
    }

    const classNames = [
      'list-view-item',
      isSelected ? 'selected' : hasSelected ? 'hidden' : ''
    ].join(' ')

    return (
      <article className={classNames} key={item.key}>
        <article className='list-view-item-title' onClick={onSelect}>
          {item.title}
        </article>
        <article className='list-view-item-content'>{item.content}</article>
      </article>
    )
  })

  return (
    <section className={`list-view${hasSelected ? ' has-selected' : ''}`}>
      {list}
    </section>
  )
}
