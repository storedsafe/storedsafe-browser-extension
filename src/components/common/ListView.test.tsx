import * as React from 'react'
import { render, unmountComponentAtNode } from 'react-dom'
import { act } from 'react-dom/test-utils'
import { ListView, ListItem } from './ListView'
import pretty from 'pretty'

/**
 * Set up and tear down container to mount tested component in.
 * */
let container: Element = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})
afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container)
  container.remove()
  container = null
})

test('<ListView />', () => {
  act(() => {
    render(<ListView items={[]} />, container)
  })

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})

test('<ListView children />', () => {
  const items: ListItem[] = [
    {
      key: 'item1',
      title: <h1>Title 1</h1>,
      content: <section>Item 1</section>
    },
    {
      key: 'item2',
      title: <h1>Title 2</h1>,
      content: <section>Item 2</section>
    }
  ]
  act(() => {
    render(<ListView items={items} />, container)
  })

  expect(pretty(container.innerHTML)).toMatchSnapshot()
})
