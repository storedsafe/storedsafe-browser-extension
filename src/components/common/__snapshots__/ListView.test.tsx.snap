// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListView /> 1`] = `"<section class=\\"list-view\\"></section>"`;

exports[`<ListView children /> 1`] = `
"<section class=\\"list-view\\">
  <article class=\\"list-view-item\\">
    <article class=\\"list-view-item-title\\">
      <h1>Title 1</h1>
    </article>
    <article class=\\"list-view-item-content\\">
      <section>Item 1</section>
    </article>
  </article>
  <article class=\\"list-view-item\\">
    <article class=\\"list-view-item-title\\">
      <h1>Title 2</h1>
    </article>
    <article class=\\"list-view-item-content\\">
      <section>Item 2</section>
    </article>
  </article>
</section>"
`;
